File: test/language/extended_members/python/ExtendedFieldInTemplateTest.py
import unittest
import zserio

from testutils import getZserioApi

class ExtendedFieldInTemplateTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "extended_members.zs").extended_field_in_template

    def testConstructorSimple(self):
        extended = self.api.ExtendedSimple()

        # always present when not read from stream
        self.assertTrue(extended.is_extended_value_present())

        # default initialized
        self.assertEqual(0, extended.value)
        self.assertEqual(0, extended.extended_value)

        extended = self.api.ExtendedSimple(42, zserio.limits.UINT32_MAX)
        self.assertTrue(extended.is_extended_value_present())
        self.assertEqual(42, extended.value)
        self.assertEqual(zserio.limits.UINT32_MAX, extended.extended_value)

    def testConstructorCompound(self):
        extended = self.api.ExtendedCompound()

        # always present when not read from stream
        self.assertTrue(extended.is_extended_value_present())

        # default initialized
        self.assertEqual(0, extended.value)
        self.assertIsNone(extended.extended_value)

        extended = self.api.ExtendedCompound(42, self.api.Compound(zserio.limits.UINT32_MAX))
        self.assertTrue(extended.is_extended_value_present())
        self.assertEqual(42, extended.value)
        self.assertEqual(zserio.limits.UINT32_MAX, extended.extended_value.field)

    def testEqSimple(self):
        extended1 = self.api.ExtendedSimple()
        extended2 = self.api.ExtendedSimple()
        self.assertEqual(extended1, extended2)

        extended1.value = 13
        self.assertNotEqual(extended1, extended2)
        extended2.value = 13
        self.assertEqual(extended1, extended2)

        extended2.extended_value = zserio.limits.UINT32_MAX
        self.assertNotEqual(extended1, extended2)
        extended1.extended_value = zserio.limits.UINT32_MAX
        self.assertEqual(extended1, extended2)

    def testEqCompound(self):
        extended1 = self.api.ExtendedCompound()
        extended2 = self.api.ExtendedCompound()
        self.assertEqual(extended1, extended2)

        extended1.value = 13
        self.assertNotEqual(extended1, extended2)
        extended2.value = 13
        self.assertEqual(extended1, extended2)

        extended2.extended_value = self.api.Compound(zserio.limits.UINT32_MAX)
        self.assertNotEqual(extended1, extended2)
        extended1.extended_value = self.api.Compound(zserio.limits.UINT32_MAX)
        self.assertEqual(extended1, extended2)

    def testHashSimple(self):
        extended1 = self.api.ExtendedSimple()
        extended2 = self.api.ExtendedSimple()
        self.assertEqual(hash(extended1), hash(extended2))

        extended1.value = 13
        self.assertNotEqual(hash(extended1), hash(extended2))
        extended2.value = 13
        self.assertEqual(hash(extended1), hash(extended2))

        extended2.extended_value = 42
        self.assertNotEqual(hash(extended1), hash(extended2))
        extended1.extended_value = 42
        self.assertEqual(hash(extended1), hash(extended2))

    def testHashCompound(self):
        extended1 = self.api.ExtendedCompound()
        extended2 = self.api.ExtendedCompound()
        self.assertEqual(hash(extended1), hash(extended2))

        extended1.value = 13
        self.assertNotEqual(hash(extended1), hash(extended2))
        extended2.value = 13
        self.assertEqual(hash(extended1), hash(extended2))

        extended2.extended_value = self.api.Compound(42)
        self.assertNotEqual(hash(extended1), hash(extended2))
        extended1.extended_value = self.api.Compound(42)
        self.assertEqual(hash(extended1), hash(extended2))

    def testBitSizeOfSimple(self):
        extended = self.api.ExtendedSimple()
        self.assertEqual(EXTENDED_BIT_SIZE, extended.bitsizeof())

    def testBitSizeOfCompound(self):
        extended = self.api.ExtendedCompound(42, self.api.Compound())
        self.assertEqual(EXTENDED_BIT_SIZE, extended.bitsizeof())

    def testInitializeOffsetsSimple(self):
        extended = self.api.ExtendedSimple()
        self.assertEqual(EXTENDED_BIT_SIZE, extended.initialize_offsets(0))

    def testInitializeOffsetsCompound(self):
        extended = self.api.ExtendedCompound(42, self.api.Compound())
        self.assertEqual(EXTENDED_BIT_SIZE, extended.initialize_offsets(0))

    def testWriteReadExtendedSimple(self):
        extended = self.api.ExtendedSimple(42, zserio.limits.UINT32_MAX)
        bitBuffer = zserio.serialize(extended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

        readExtended = zserio.deserialize(self.api.ExtendedSimple, bitBuffer)
        self.assertTrue(readExtended.is_extended_value_present())
        self.assertEqual(extended, readExtended)

    def testWriteReadExtendedCompound(self):
        extended = self.api.ExtendedCompound(42, self.api.Compound(zserio.limits.UINT32_MAX))
        bitBuffer = zserio.serialize(extended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

        readExtended = zserio.deserialize(self.api.ExtendedCompound, bitBuffer)
        self.assertTrue(readExtended.is_extended_value_present())
        self.assertEqual(extended, readExtended)

    def testWriteOriginalReadExtendedSimple(self):
        original = self.api.Original(42)
        bitBuffer = zserio.serialize(original)
        readExtended = zserio.deserialize(self.api.ExtendedSimple, bitBuffer)
        self.assertFalse(readExtended.is_extended_value_present())

        # extended value is default constructed
        self.assertEqual(0, readExtended.extended_value)

        # bit size as original
        self.assertEqual(ORIGINAL_BIT_SIZE, readExtended.bitsizeof())

        # initialize offsets as original
        self.assertEqual(ORIGINAL_BIT_SIZE, readExtended.initialize_offsets(0))

        # write as original
        bitBuffer = zserio.serialize(readExtended)
        self.assertEqual(ORIGINAL_BIT_SIZE, bitBuffer.bitsize)

        # read original again
        readOriginal = zserio.deserialize(self.api.Original, bitBuffer)
        self.assertEqual(original, readOriginal)

        # setter makes the value present!
        readExtended.extended_value = zserio.limits.UINT32_MAX
        self.assertTrue(readExtended.is_extended_value_present())

        # bit size as extended
        self.assertEqual(EXTENDED_BIT_SIZE, readExtended.bitsizeof())

        # initialize offsets as extended
        self.assertEqual(EXTENDED_BIT_SIZE, readExtended.initialize_offsets(0))

        # writes as extended
        bitBuffer = zserio.serialize(readExtended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

    def testWriteOriginalReadExtendedCompound(self):
        original = self.api.Original(42)
        bitBuffer = zserio.serialize(original)
        readExtended = zserio.deserialize(self.api.ExtendedCompound, bitBuffer)
        self.assertFalse(readExtended.is_extended_value_present())

        # extended value is None
        self.assertIsNone(readExtended.extended_value)

        # bit size as original
        self.assertEqual(ORIGINAL_BIT_SIZE, readExtended.bitsizeof())

        # initialize offsets as original
        self.assertEqual(ORIGINAL_BIT_SIZE, readExtended.initialize_offsets(0))

        # write as original
        bitBuffer = zserio.serialize(readExtended)
        self.assertEqual(ORIGINAL_BIT_SIZE, bitBuffer.bitsize)

        # read original again
        readOriginal = zserio.deserialize(self.api.Original, bitBuffer)
        self.assertEqual(original, readOriginal)

        # setter makes the value present!
        readExtended.extended_value = self.api.Compound(zserio.limits.UINT32_MAX)
        self.assertTrue(readExtended.is_extended_value_present())

        # bit size as extended
        self.assertEqual(EXTENDED_BIT_SIZE, readExtended.bitsizeof())

        # initialize offsets as extended
        self.assertEqual(EXTENDED_BIT_SIZE, readExtended.initialize_offsets(0))

        # writes as extended
        bitBuffer = zserio.serialize(readExtended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

    def testWriteExtendedSimpleReadOriginal(self):
        extended = self.api.ExtendedSimple(42, zserio.limits.UINT32_MAX)
        bitBuffer = zserio.serialize(extended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

        reader = zserio.BitStreamReader.from_bitbuffer(bitBuffer)
        readOriginal = self.api.Original.from_reader(reader)
        self.assertEqual(extended.value, readOriginal.value)
        self.assertEqual(ORIGINAL_BIT_SIZE, reader.bitposition)

    def testWriteExtendedCompoundReadOriginal(self):
        extended = self.api.ExtendedCompound(42, self.api.Compound(zserio.limits.UINT32_MAX))
        bitBuffer = zserio.serialize(extended)
        self.assertEqual(EXTENDED_BIT_SIZE, bitBuffer.bitsize)

        reader = zserio.BitStreamReader.from_bitbuffer(bitBuffer)
        readOriginal = self.api.Original.from_reader(reader)
        self.assertEqual(extended.value, readOriginal.value)
        self.assertEqual(ORIGINAL_BIT_SIZE, reader.bitposition)

ORIGINAL_BIT_SIZE = 4 * 8
EXTENDED_BIT_SIZE = ORIGINAL_BIT_SIZE + 4 * 8

File: test/language/extended_members/cpp/ExtendedFieldInTemplateTest.cpp
#include "extended_members/extended_field_in_template/ExtendedCompound.h"
#include "extended_members/extended_field_in_template/ExtendedSimple.h"
#include "extended_members/extended_field_in_template/Original.h"
#include "gtest/gtest.h"
#include "zserio/SerializeUtil.h"

namespace extended_members
{
namespace extended_field_in_template
{

using allocator_type = ExtendedCompound::allocator_type;

class ExtendedFieldInTemplateTest : public ::testing::Test
{
protected:
    template <typename EXTENDED>
    void checkCopyAndMove(const EXTENDED& extended, bool expectedIsPresent)
    {
        auto copiedExtended(extended);
        ASSERT_EQ(expectedIsPresent, copiedExtended.isExtendedValuePresent());
        ASSERT_EQ(extended, copiedExtended);

        auto movedExtended(std::move(copiedExtended));
        ASSERT_EQ(expectedIsPresent, movedExtended.isExtendedValuePresent());
        ASSERT_EQ(extended, movedExtended);

        EXTENDED copiedWithPropagateAllocatorExtended(zserio::PropagateAllocator, extended, allocator_type());
        ASSERT_EQ(expectedIsPresent, copiedWithPropagateAllocatorExtended.isExtendedValuePresent());
        ASSERT_EQ(extended, copiedWithPropagateAllocatorExtended);

        EXTENDED copyAssignedExtended;
        copyAssignedExtended = extended;
        ASSERT_EQ(expectedIsPresent, copyAssignedExtended.isExtendedValuePresent());
        ASSERT_EQ(extended, copyAssignedExtended);

        EXTENDED moveAssignedExtended;
        moveAssignedExtended = copyAssignedExtended;
        ASSERT_EQ(expectedIsPresent, moveAssignedExtended.isExtendedValuePresent());
        ASSERT_EQ(extended, moveAssignedExtended);
    }

    static const size_t ORIGINAL_BIT_SIZE;
    static const size_t EXTENDED_BIT_SIZE;
};

const size_t ExtendedFieldInTemplateTest::ORIGINAL_BIT_SIZE = 4 * 8;
const size_t ExtendedFieldInTemplateTest::EXTENDED_BIT_SIZE = ORIGINAL_BIT_SIZE + 4 * 8;

TEST_F(ExtendedFieldInTemplateTest, defaultConstructorSimple)
{
    ExtendedSimple extended;

    // always present when not read from stream
    ASSERT_TRUE(extended.isExtendedValuePresent());

    // default initialized
    ASSERT_EQ(0, extended.getValue());
    ASSERT_EQ(0, extended.getExtendedValue());
}

TEST_F(ExtendedFieldInTemplateTest, defaultConstructorCompound)
{
    ExtendedCompound extended;

    // always present when not read from stream
    ASSERT_TRUE(extended.isExtendedValuePresent());

    // default initialized
    ASSERT_EQ(0, extended.getValue());
    ASSERT_EQ(Compound(), extended.getExtendedValue());
}

TEST_F(ExtendedFieldInTemplateTest, fieldConstructorSimple)
{
    ExtendedSimple extended(42, UINT32_MAX);
    ASSERT_TRUE(extended.isExtendedValuePresent());

    ASSERT_EQ(42, extended.getValue());
    ASSERT_EQ(UINT32_MAX, extended.getExtendedValue());
}

TEST_F(ExtendedFieldInTemplateTest, fieldConstructorCompound)
{
    ExtendedCompound extended(42, Compound(UINT32_MAX));
    ASSERT_TRUE(extended.isExtendedValuePresent());

    ASSERT_EQ(42, extended.getValue());
    ASSERT_EQ(UINT32_MAX, extended.getExtendedValue().getField());
}

TEST_F(ExtendedFieldInTemplateTest, operatorEqualitySimple)
{
    ExtendedSimple extended1;
    ExtendedSimple extended2;
    ASSERT_EQ(extended1, extended2);

    extended1.setValue(13);
    ASSERT_FALSE(extended1 == extended2);
    extended2.setValue(13);
    ASSERT_EQ(extended1, extended2);

    extended2.setExtendedValue(UINT32_MAX);
    ASSERT_FALSE(extended1 == extended2);
    extended1.setExtendedValue(UINT32_MAX);
    ASSERT_EQ(extended1, extended2);
}

TEST_F(ExtendedFieldInTemplateTest, operatorLessThanSimple)
{
    ExtendedSimple extended1;
    ExtendedSimple extended2;
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended1.setValue(13);
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_TRUE(extended2 < extended1);

    extended2.setValue(13);
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended2.setExtendedValue(UINT32_MAX);
    ASSERT_TRUE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended1.setExtendedValue(UINT32_MAX);
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);
}

TEST_F(ExtendedFieldInTemplateTest, operatorEqualityCompound)
{
    ExtendedCompound extended1;
    ExtendedCompound extended2;
    ASSERT_EQ(extended1, extended2);

    extended1.setValue(13);
    ASSERT_FALSE(extended1 == extended2);
    extended2.setValue(13);
    ASSERT_EQ(extended1, extended2);

    extended2.setExtendedValue(Compound(UINT32_MAX));
    ASSERT_FALSE(extended1 == extended2);
    extended1.setExtendedValue(Compound(UINT32_MAX));
    ASSERT_EQ(extended1, extended2);
}

TEST_F(ExtendedFieldInTemplateTest, operatorLessThanCompound)
{
    ExtendedCompound extended1;
    ExtendedCompound extended2;
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended1.setValue(13);
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_TRUE(extended2 < extended1);

    extended2.setValue(13);
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended2.setExtendedValue(Compound(UINT32_MAX));
    ASSERT_TRUE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);

    extended1.setExtendedValue(Compound(UINT32_MAX));
    ASSERT_FALSE(extended1 < extended2);
    ASSERT_FALSE(extended2 < extended1);
}

TEST_F(ExtendedFieldInTemplateTest, hashCodeSimple)
{
    ExtendedSimple extended1;
    ExtendedSimple extended2;
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());

    extended1.setValue(13);
    ASSERT_NE(extended1.hashCode(), extended2.hashCode());
    extended2.setValue(13);
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());

    extended2.setExtendedValue(42);
    ASSERT_NE(extended1.hashCode(), extended2.hashCode());
    extended1.setExtendedValue(42);
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());
}

TEST_F(ExtendedFieldInTemplateTest, hashCodeCompound)
{
    ExtendedCompound extended1;
    ExtendedCompound extended2;
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());

    extended1.setValue(13);
    ASSERT_NE(extended1.hashCode(), extended2.hashCode());
    extended2.setValue(13);
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());

    extended2.setExtendedValue(Compound(42));
    ASSERT_NE(extended1.hashCode(), extended2.hashCode());
    extended1.setExtendedValue(Compound(42));
    ASSERT_EQ(extended1.hashCode(), extended2.hashCode());
}

TEST_F(ExtendedFieldInTemplateTest, bitSizeOfSimple)
{
    ExtendedSimple extended;
    ASSERT_EQ(EXTENDED_BIT_SIZE, extended.bitSizeOf());
}

TEST_F(ExtendedFieldInTemplateTest, bitSizeOfCompound)
{
    ExtendedCompound extended;
    ASSERT_EQ(EXTENDED_BIT_SIZE, extended.bitSizeOf());
}

TEST_F(ExtendedFieldInTemplateTest, initializeOffsetsSimple)
{
    ExtendedSimple extended;
    ASSERT_EQ(EXTENDED_BIT_SIZE, extended.initializeOffsets(0));
}

TEST_F(ExtendedFieldInTemplateTest, initializeOffsetsCompound)
{
    ExtendedCompound extended;
    ASSERT_EQ(EXTENDED_BIT_SIZE, extended.initializeOffsets(0));
}

TEST_F(ExtendedFieldInTemplateTest, writeReadExtendedSimple)
{
    ExtendedSimple extended(42, UINT32_MAX);
    auto bitBuffer = zserio::serialize(extended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    auto readExtended = zserio::deserialize<ExtendedSimple>(bitBuffer);
    ASSERT_TRUE(readExtended.isExtendedValuePresent());
    ASSERT_EQ(extended, readExtended);

    checkCopyAndMove(extended, true);
}

TEST_F(ExtendedFieldInTemplateTest, writeReadExtendedCompound)
{
    ExtendedCompound extended(42, Compound(UINT32_MAX));
    auto bitBuffer = zserio::serialize(extended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    auto readExtended = zserio::deserialize<ExtendedCompound>(bitBuffer);
    ASSERT_TRUE(readExtended.isExtendedValuePresent());
    ASSERT_EQ(extended, readExtended);

    checkCopyAndMove(extended, true);
}

TEST_F(ExtendedFieldInTemplateTest, writeOriginalReadExtendedSimple)
{
    Original original(42);
    auto bitBuffer = zserio::serialize(original);
    auto readExtended = zserio::deserialize<ExtendedSimple>(bitBuffer);
    ASSERT_FALSE(readExtended.isExtendedValuePresent());

    // extended value is default constructed
    ASSERT_EQ(0, readExtended.getExtendedValue());

    // bit size as original
    ASSERT_EQ(ORIGINAL_BIT_SIZE, readExtended.bitSizeOf());

    // initialize offsets as original
    ASSERT_EQ(ORIGINAL_BIT_SIZE, readExtended.initializeOffsets(0));

    // write as original
    bitBuffer = zserio::serialize(readExtended);
    ASSERT_EQ(ORIGINAL_BIT_SIZE, bitBuffer.getBitSize());

    // read original again
    auto readOriginal = zserio::deserialize<Original>(bitBuffer);
    ASSERT_EQ(original, readOriginal);

    checkCopyAndMove(readExtended, false);

    // setter makes the value present!
    readExtended.setExtendedValue(UINT32_MAX);
    ASSERT_TRUE(readExtended.isExtendedValuePresent());

    // bit size as extended
    ASSERT_EQ(EXTENDED_BIT_SIZE, readExtended.bitSizeOf());

    // initialize offsets as extended
    ASSERT_EQ(EXTENDED_BIT_SIZE, readExtended.initializeOffsets(0));

    // writes as extended
    bitBuffer = zserio::serialize(readExtended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    checkCopyAndMove(readExtended, true);
}

TEST_F(ExtendedFieldInTemplateTest, writeOriginalReadExtendedCompound)
{
    Original original(42);
    auto bitBuffer = zserio::serialize(original);
    auto readExtended = zserio::deserialize<ExtendedCompound>(bitBuffer);
    ASSERT_FALSE(readExtended.isExtendedValuePresent());

    // extended value is default constructed
    ASSERT_EQ(Compound(), readExtended.getExtendedValue());

    // bit size as original
    ASSERT_EQ(ORIGINAL_BIT_SIZE, readExtended.bitSizeOf());

    // initialize offsets as original
    ASSERT_EQ(ORIGINAL_BIT_SIZE, readExtended.initializeOffsets(0));

    // write as original
    bitBuffer = zserio::serialize(readExtended);
    ASSERT_EQ(ORIGINAL_BIT_SIZE, bitBuffer.getBitSize());

    // read original again
    auto readOriginal = zserio::deserialize<Original>(bitBuffer);
    ASSERT_EQ(original, readOriginal);

    checkCopyAndMove(readExtended, false);

    // setter makes the value present!
    readExtended.setExtendedValue(Compound(UINT32_MAX));
    ASSERT_TRUE(readExtended.isExtendedValuePresent());

    // bit size as extended
    ASSERT_EQ(EXTENDED_BIT_SIZE, readExtended.bitSizeOf());

    // initialize offsets as extended
    ASSERT_EQ(EXTENDED_BIT_SIZE, readExtended.initializeOffsets(0));

    // writes as extended
    bitBuffer = zserio::serialize(readExtended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    checkCopyAndMove(readExtended, true);
}

TEST_F(ExtendedFieldInTemplateTest, writeExtendedSimpleReadOriginal)
{
    ExtendedSimple extended(42, UINT32_MAX);
    auto bitBuffer = zserio::serialize(extended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    zserio::BitStreamReader reader(bitBuffer);
    Original readOriginal(reader);
    ASSERT_EQ(extended.getValue(), readOriginal.getValue());
    ASSERT_EQ(ORIGINAL_BIT_SIZE, reader.getBitPosition());
}

TEST_F(ExtendedFieldInTemplateTest, writeExtendedCompoundReadOriginal)
{
    ExtendedCompound extended(42, Compound(UINT32_MAX));
    auto bitBuffer = zserio::serialize(extended);
    ASSERT_EQ(EXTENDED_BIT_SIZE, bitBuffer.getBitSize());

    zserio::BitStreamReader reader(bitBuffer);
    Original readOriginal(reader);
    ASSERT_EQ(extended.getValue(), readOriginal.getValue());
    ASSERT_EQ(ORIGINAL_BIT_SIZE, reader.getBitPosition());
}

} // namespace extended_field_in_template
} // namespace extended_members
